crates/typst/src/export/svg.rs

```diff
@@ -408,6 +617,107 @@ impl SVGRenderer {
 
         self.xml.end_element();
     }
+
+    /// Write the raw gradients (without transform) to the SVG file.
+    fn write_gradients(&mut self) {
+        if self.gradients.is_empty() {
+            return;
+        }
+
+        self.xml.start_element("defs");
+        self.xml.write_attribute("id", "gradients");
+
+        for (id, (gradient, ratio)) in self.gradients.iter() {
+            match &gradient {
+                Gradient::Linear(linear) => {
+                    self.xml.start_element("linearGradient");
+                    self.xml.write_attribute("id", &id);
+                    self.xml.write_attribute("spreadMethod", "pad");
+                    self.xml.write_attribute("gradientUnits", "userSpaceOnUse");
+
+                    let angle = linear.angle.correct_aspect_ratio(*ratio);
+                    let (sin, cos) = angle.sin_cos();
+                    let length = sin.abs() + cos.abs();
+                    let (x1, y1, x2, y2) = match angle.quadrant() {
+                        Quadrant::First => (0.0, 0.0, cos * length, sin * length),
+                        Quadrant::Second => (1.0, 0.0, cos * length + 1.0, sin * length),
+                        Quadrant::Third => {
+                            (1.0, 1.0, cos * length + 1.0, sin * length + 1.0)
+                        }
+                        Quadrant::Fourth => (0.0, 1.0, cos * length, sin * length + 1.0),
+                    };
+
+                    self.xml.write_attribute("x1", &x1);
+                    self.xml.write_attribute("y1", &y1);
+                    self.xml.write_attribute("x2", &x2);
+                    self.xml.write_attribute("y2", &y2);
+
+                    for window in linear.stops.windows(2) {
+                        let (start_c, start_t) = window[0];
+                        let (end_c, end_t) = window[1];
+
+                        // Generate 32 stops between the two stops.
+                        // This is a workaround for a bug in many readers:
+                        // They tend to just ignore the color space of the gradient.
+                        for i in 0..=32 {
+                            let t0 = i as f64 / 32.0;
+                            let t = start_t + (end_t - start_t) * t0;
+                            let c = Color::mix(
+                                vec![
+                                    WeightedColor::new(start_c, 1.0 - t0),
+                                    WeightedColor::new(end_c, t0),
+                                ],
```

Comment:
This allocates potentially quite a lot. Either we have a separate mix function with slices or we could at least preallocate and reuse the same vector.